### PR TITLE
Migrate to tauri v2

### PR DESCRIPTION
**Pull request by pierremtb**

### CODE REVIEW

**Comment by lf94** [comment]
What else do we need to do to make this a reality? After my current work I wouldn't mind pushing this through next.

**Comment by paultag** [comment]
@lf94 I spent some time getting this to play with WebRTC; some notes:

tl;dr: This wasn't an easy fix. There's continued debugging to be done

1) I was able to get tauri 2 and WebRTC enabled webkit2gtk to work without issue. The hardest bit was a few rebuilds; very low drama. My old notes stand on that part.

2) Once I had a WebRTC enabled webkit2gtk, I was trying to get it to talk webrtc. It was posible, but very flaky, and I was having trouble pinning it down. I got exactly one WebRTC example application to work, but the brand name code (I think I tried jitsi?) didn't.

3) KCMA failed during negotiation, leading to https://github.com/webrtc-rs/webrtc/issues/533#issuecomment-1960771329 -- I don't think that issue I filed is the root cause after more digging, but I can't be sure. I think that's "just" what happens if we fail to WebRTC handshake (SDP/codec issue? Not sure, haven't pinned it yet). I can't be sure, so I haven't followed up on that bug yet.

There's work to be done here, but it's not as easy as a rebuild/run cycle unfortunately. I did spend about a day digging around on this one, it's close but not there yet. I was going to re-try after I finished my overhaul of the WebRTC backend code so I can more easily put in debugging stubs to figure out what specifically is going on here.

**Comment by lf94** [comment]
Excellent, thanks @paultag ! My take away from this is webkit2gtk's WebRTC support continues to be pretty bad then.

**Comment by lf94** [comment]
I'm looking to pivot to this as my next 1.0 unblocker task.

There are two paths:

1. We fork the necessary libraries and make any changes we need to make Tauri 2 work for us. This is quite an undertaking.
2. We migrate to Electron, another somewhat heavy task.

There are benefits to both. Both involve migrations. But only one involves becoming responsible for new, very large codebases, and that's if we stick to Tauri 2.

Electron has the additional benefit of being a consistent platform for us across **all operating systems**.

I propose we migrate to Electron.

@jessfraz ?

**Comment by jessfraz** [comment]
id say this timebox the electron thing and if it starts to open even more rabbit holes and mess drop it and do the tauri stuff but @paultag knows better than me the limitations we are hitting with tauri and if its a huge pain in the ass to fix

**Comment by pierremtb** [comment]
Potential lead on what's going wrong with Win release builds: https://github.com/tauri-apps/tauri/issues/9045

**Comment by franknoirot** [comment]
Wow this is excellent.

Do you happen to know why they got rid of the recursive option? What a bummer of a loss

File: src-tauri/capabilities/migrated.json
Comment (franknoirot) [inline src-tauri/capabilities/migrated.json:1]: Are these used, or being kept here for posterity? I don't see how they would be loaded, maybe a magic file location?

File: src/components/FileMachineProvider.tsx [hunk 1]
```diff
@@ -99,12 +92,11 @@ export const FileMachineProvider = ({
...
-          context.selectedDirectory.path + sep + oldName,
-          context.selectedDirectory.path +
-            sep +
-            name +
-            (name.endsWith(FILE_EXT) || isDir ? '' : FILE_EXT)
+        await rename(
+          await join(context.selectedDirectory.path, oldName),
+          (await join(context.selectedDirectory.path, name)) +
+            (name.endsWith(FILE_EXT) || isDir ? '' : FILE_EXT),
+          {}
```
Comment (franknoirot) [inline src/components/FileMachineProvider.tsx:99]: Nice this much more readable now

File: src/lib/types.ts [hunk 2]
```diff
@@ -14,12 +13,25 @@ export type FileLoaderData = {
...
+  path: string
+  /**
+   * Name of the directory/file
+   * can be null if the path terminates with `..`
+   */
+  name?: string
+  /** Children of this entry if it's a directory; null otherwise */
+  children?: FileEntry[]
+}
+
```
Comment (franknoirot) [inline src/lib/types.ts:34]: Is there a better type I should use from `@tauri-apps/plugin-fs`? I'd rather we use their type if they have one exposed, otherwise couldn't we have a mismatch between our expected shape of the data and the actually-returned data from their read function. I guess it's not a concern since we're shimming our own `read_dir_recursive` function from V1 everywhere?

**Review by franknoirot — COMMENTED** [review]
This is good work @pierremtb! I just have a couple of questions and I want to allow other teammates a chance to approve since it's a big update. I'll get around to some testing of the built versions of this today or Monday

**Review by adamchalmers — APPROVED** [review]

File: src-tauri/capabilities/migrated.json
Comment (pierremtb) [reply src-tauri/capabilities/migrated.json:1]: They're used. I don't know why the name is migrated.json. That was the auto porting script's making. We should rename it to something less confusing

File: package.json [hunk 3]
```diff
@@ -16,7 +16,13 @@
     "@open-rpc/client-js": "^1.8.1",
     "@react-hook/resize-observer": "^1.2.6",
     "@replit/codemirror-interact": "^6.3.0",
-    "@tauri-apps/api": "^1.5.3",
+    "@sentry/react": "^7.77.0",
```
[the remark this answers is not in the document]
Comment (pierremtb) [reply package.json:19]: My bad! Will remove

File: src-tauri/capabilities/migrated.json
Comment (pierremtb) [reply src-tauri/capabilities/migrated.json:1]: Renamed to `desktop.json` which looks standard from what I saw at https://beta.tauri.app/references/v2/acl/

File: src/lib/types.ts (hunk 2 above)
Comment (pierremtb) [reply src/lib/types.ts:34]: @franknoirot There is a new type yes, but it's not recursive-able. So unless we find a way to rewrite the filesystem reading logic we have to use the non-recursive `readDir` we can't really use it I think

It's `DiskEntry`, defined at https://github.com/tauri-apps/plugins-workspace/blob/v2/plugins/fs/guest-js/index.ts#L625-L641

**Comment by pierremtb** [comment]
> Wow this is excellent.
> 
> Do you happen to know why they got rid of the recursive option? What a bummer of a loss

@franknoirot No I can't find anything that points to why

**Review by Irev-Dev — APPROVED** [review]

**Comment by pierremtb** [comment]
Merging after extensive manual testing of the Windows and macOS builds.
